packages/next/src/shared/lib/router/action-queue.ts

```diff
@@ -0,0 +1,150 @@
+
+export type DispatchStatePromise = React.Dispatch<ReducerState>
+
+export type AppRouterActionQueue = {
+  state: AppRouterState | null
+  devToolsInstance?: ReduxDevToolsInstance
+  dispatch: (payload: ReducerActions, setState: DispatchStatePromise) => void
+  action: (state: AppRouterState, action: ReducerActions) => ReducerState
+  pending: ActionQueueNode | null
+  last: ActionQueueNode | null
+}
+
+export type ActionQueueNode = {
+  payload: ReducerActions
+  next: ActionQueueNode | null
+  resolve: (value: PromiseLike<AppRouterState> | AppRouterState) => void
+  reject: (err: Error) => void
+}
+
+export const ActionQueueContext =
+  React.createContext<AppRouterActionQueue | null>(null)
+
+function runRemainingActions(
+  actionQueue: AppRouterActionQueue,
+  setState: DispatchStatePromise
+) {
+  if (actionQueue.pending !== null) {
+    actionQueue.pending = actionQueue.pending.next
+    if (actionQueue.pending !== null) {
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      runAction({
+        actionQueue,
+        action: actionQueue.pending,
+        setState,
+      })
+    }
+  }
+}
+
+async function runAction({
+  actionQueue,
+  action,
+  setState,
+}: {
+  actionQueue: AppRouterActionQueue
+  action: ActionQueueNode
+  setState: DispatchStatePromise
+}) {
+  const prevState = actionQueue.state
+  if (!prevState) {
+    // This shouldn't happen as the state is initialized in the dispatcher if it's not set
+    throw new Error('Invariant: Router state not initialized')
+  }
+
+  const payload = action.payload
+  const actionResult = actionQueue.action(prevState, payload)
+
+  function handleResult(nextState: AppRouterState) {
+    actionQueue.state = nextState
+
+    if (actionQueue.devToolsInstance) {
+      actionQueue.devToolsInstance.send(payload, nextState)
+    }
+
+    runRemainingActions(actionQueue, setState)
+    action.resolve(nextState)
+  }
+
+  // if the action is a promise, set up a callback to resolve it
+  if (actionResult instanceof Promise) {
+    actionResult.then(handleResult, (err) => {
+      runRemainingActions(actionQueue, setState)
+      action.reject(err)
+    })
+  } else {
+    handleResult(actionResult)
+  }
+}
+
+function dispatchAction(
+  actionQueue: AppRouterActionQueue,
+  payload: ReducerActions,
+  setState: DispatchStatePromise
+) {
+  const {
+    promise: deferredPromise,
+    resolve,
+    reject,
+  } = Promise.withResolvers<AppRouterState>()
+
+  const newAction: ActionQueueNode = {
+    payload,
+    next: null,
+    resolve,
+    reject,
+  }
+
+  startTransition(() => {
+    // we immediately notify React of the pending promise -- the resolver is attached to the action node
+    // and will be called when the associated action promise resolves
+    setState(deferredPromise)
+  })
+
+  // Check if the queue is empty or if the action is a navigation action
+  if (actionQueue.pending === null || payload.type === 'navigate') {
```

Comment:
Can you fill in the context for me here? What's a "navigate" action and why is it a special case? This is fishy to me because it's still marked as a transition.

Part of the rationale of hoisting out the queue is that router actions should be able to be sequentially processed. We shouldn't be running multiple actions concurrently because it will lead to data races. So besides blocking, another option is to abort whatever action is currently pending.

Reply:
This was to address a scenario where if you fired off a long-running server action, navigations were blocked until the action runs to completion.

I'm also ok with the aborting strategy, but I was assuming people would still expect their action to run to completion, even if the result of it was discarded. 

Here's the test for that scenario: https://github.com/vercel/next.js/blob/canary/test/e2e/app-dir/actions/app-action.test.ts#L271-L289

Reply:
Ah ok, that sounds good to me! Will update